README.md
# fluent-plugin-gcs
[![Gem Version](https://badge.fury.io/rb/fluent-plugin-gcs.svg)](https://badge.fury.io/rb/fluent-plugin-gcs) [![Test](https://github.com/daichirata/fluent-plugin-gcs/actions/workflows/test.yaml/badge.svg)](https://github.com/daichirata/fluent-plugin-gcs/actions/workflows/test.yaml) [![Code Climate](https://codeclimate.com/github/daichirata/fluent-plugin-gcs/badges/gpa.svg)](https://codeclimate.com/github/daichirata/fluent-plugin-gcs)

Google Cloud Storage output plugin for [Fluentd](https://github.com/fluent/fluentd).

## Requirements

| fluent-plugin-gcs  | fluentd    | ruby   |
|--------------------|------------|--------|
| >= 0.4.0           | >= v0.14.0 | >= 2.4 |
|  < 0.4.0           | >= v0.12.0 | >= 1.9 |

## Installation

``` shell
$ gem install fluent-plugin-gcs -v "~> 0.3" --no-document # for fluentd v0.12 or later
$ gem install fluent-plugin-gcs -v "0.4.0" --no-document # for fluentd v0.14 or later
```

## Examples

### For v0.14 style

```
<match pattern>
  @type gcs

  project YOUR_PROJECT
  keyfile YOUR_KEYFILE_PATH
  bucket YOUR_GCS_BUCKET_NAME
  object_key_format %{path}%{time_slice}_%{index}.%{file_extension}
  path logs/${tag}/%Y/%m/%d/

  # if you want to use ${tag} or %Y/%m/%d/ like syntax in path / object_key_format,
  # need to specify tag for ${tag} and time for %Y/%m/%d in <buffer> argument.
  <buffer tag,time>
    @type file
    path /var/log/fluent/gcs
    timekey 1h # 1 hour partition
    timekey_wait 10m
    timekey_use_utc true # use utc
  </buffer>

  <format>
    @type json
  </format>
</match>
```

### For v0.12 style

```
<match pattern>
  @type gcs

  project YOUR_PROJECT
  keyfile YOUR_KEYFILE_PATH
  bucket YOUR_GCS_BUCKET_NAME
  object_key_format %{path}%{time_slice}_%{index}.%{file_extension}
  path logs/
  buffer_path /var/log/fluent/gcs

  time_slice_format %Y%m%d-%H
  time_slice_wait 10m
  utc
</match>
```

## Configuration

### Authentication

You can provide the project and credential information to connect to the Storage
service, or if you are running on Google Compute Engine this configuration is taken care of for you.

**project**

Project identifier for GCS. Project are discovered in the following order:
* Specify project in `project`
* Discover project in environment variables `STORAGE_PROJECT`, `GOOGLE_CLOUD_PROJECT`, `GCLOUD_PROJECT`
* Discover GCE credentials

**keyfile**

Path of GCS service account credentials JSON file. Credentials are discovered in the following order:
* Specify credentials path in `keyfile`
* Discover credentials path in environment variables `GOOGLE_CLOUD_KEYFILE`, `GCLOUD_KEYFILE`
* Discover credentials JSON in environment variables `GOOGLE_CLOUD_KEYFILE_JSON`, `GCLOUD_KEYFILE_JSON`
* Discover credentials file in the Cloud SDK's path
* Discover GCE credentials

**client_retries**

Number of times to retry requests on server error.

**client_timeout**

Default timeout to use in requests.

**bucket (*required)**

GCS bucket name.

**store_as**

Archive format on GCS. You can use serveral format:

* gzip (default)
* json
* text

**path**

path prefix of the files on GCS. Default is "" (no prefix).

**object_key_format**

The format of GCS object keys. You can use several built-in variables:

* %{path}
* %{time_slice}
* %{index}
* %{file_extension}
* %{uuid_flush}
* %{hex_random}
* %{hostname}

to decide keys dynamically.

* `%{path}` is exactly the value of `path` configured in the configuration file. E.g., "logs/" in the example configuration above.
* `%{time_slice}` is the time-slice in text that are formatted with `time_slice_format`.
* `%{index}` is the sequential number starts from 0, increments when multiple files are uploaded to GCS in the same time slice.
* `%{file_extention}` is changed by the value of `store_as`.
  * gzip - gz
  * json - json
  * text - txt
* `%{uuid_flush}` a uuid that is replaced everytime the buffer will be flushed
* `%{hex_random}` a random hex string that is replaced for each buffer chunk, not assured to be unique. You can configure the length of string with a `hex_random_length` parameter (Default: 4).
* `%{hostname}` is set to the standard host name of the system of the running server.

The default format is `%{path}%{time_slice}_%{index}.%{file_extension}`.

**hex_random_length**

The length of `%{hex_random}` placeholder.

**transcoding**

Enable the decompressive form of transcoding.

See also [Transcoding of gzip-compressed files](https://cloud.google.com/storage/docs/transcoding).

**format**

Change one line format in the GCS object. You can use serveral format:

* out_file (default)
* json
* ltsv
* single_value

See also [official Formatter article](http://docs.fluentd.org/articles/formatter-plugin-overview).

**auto_create_bucket**

Create GCS bucket if it does not exists. Default is true.

**acl**

Permission for the object in GCS. Acceptable values are:

* `auth_read`       - File owner gets OWNER access, and allAuthenticatedUsers get READER access.
* `owner_full`      - File owner gets OWNER access, and project team owners get OWNER access.
* `owner_read`      - File owner gets OWNER access, and project team owners get READER access.
* `private`         - File owner gets OWNER access.
* `project_private` - File owner gets OWNER access, and project team members get access according to their roles.
* `public_read`     - File owner gets OWNER access, and allUsers get READER access.

Default is nil (bucket default object ACL). See also [official document](https://cloud.google.com/storage/docs/access-control/lists).

**storage_class**

Storage class of the file. Acceptable values are:

* `dra`            - Durable Reduced Availability
* `nearline`       - Nearline Storage
* `coldline`       - Coldline Storage
* `multi_regional` - Multi-Regional Storage
* `regional`       - Regional Storage
* `standard`       - Standard Storage

Default is nil. See also [official document](https://cloud.google.com/storage/docs/storage-classes).

**encryption_key**

You can also choose to provide your own AES-256 key for server-side encryption. See also [Customer-supplied encryption keys](https://cloud.google.com/storage/docs/encryption#customer-supplied).

`encryption_key_sha256` will be calculated using encryption_key.

**overwrite**

Overwrite already existing path. Default is false, which raises an error
if a GCS object of the same path already exists, or increment the
`%{index}` placeholder until finding an absent path.

**buffer_path (*required)**

path prefix of the files to buffer logs.

**time_slice_format**

Format of the time used as the file name. Default is '%Y%m%d'. Use
'%Y%m%d%H' to split files hourly.

**time_slice_wait**

The time to wait old logs. Default is 10 minutes. Specify larger value if
old logs may reache.

**localtime**

Use Local time instead of UTC.

**utc**

Use UTC instead of local time.

And see [official Time Sliced Output article](http://docs.fluentd.org/articles/output-plugin-overview#time-sliced-output-parameters)

**blind_write**

Doesn't check if an object exists in GCS before writing. Default is false.

Allows to avoid granting of `storage.objects.get` permission.

Warning! If the object exists and `storage.objects.delete` permission is not
granted, it will result in an unrecoverable error. Usage of `%{hex_random}` is
recommended.

### ObjectMetadata

User provided web-safe keys and arbitrary string values that will returned with requests for the file as "x-goog-meta-" response headers.

```
<match *>
  @type gcs

  <object_metadata>
    key KEY_DATA_1
    value VALUE_DATA_1
  </object_metadata>

  <object_metadata>
    key KEY_DATA_2
    value VALUE_DATA_2
  </object_metadata>
</match>
```


CHANGELOG.md
## [Unreleased]

New features / Enhancements

## [0.4.2] - 2022/08/16

Bug fixes

- [Fix automatic conversion from a hash to keyword arguments](https://github.com/daichirata/fluent-plugin-gcs/pull/22)

## [0.4.1] - 2020/04/17

New features
- [Support blind write to GSC](https://github.com/daichirata/fluent-plugin-gcs/pull/14)

## [0.4.0] - 2019/04/01

New features / Enhancements

- [Support v0.14 (by @cosmo0920)](https://github.com/daichirata/fluent-plugin-gcs/pull/6)

## [0.3.0] - 2017/02/28

New features / Enhancements

- [Add support for setting a File's storage_class on file creation](https://github.com/daichirata/fluent-plugin-gcs/pull/4)
  - see also https://cloud.google.com/storage/docs/storage-classes

## [0.2.0] - 2017/01/16

Bug fixes

- [Remove encryption_key_sha256 parameter.](https://github.com/daichirata/fluent-plugin-gcs/pull/2)
  - see also. https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-storage/CHANGELOG.md#0230--2016-12-8

## [0.1.1] - 2016/11/28

New features / Enhancements

- Add support for `%{hostname}` of object_key_format

[Unreleased]: https://github.com/daichirata/fluent-plugin-gcs/compare/v0.4.0...HEAD
[0.4.0]: https://github.com/daichirata/fluent-plugin-gcs/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/daichirata/fluent-plugin-gcs/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/daichirata/fluent-plugin-gcs/compare/v0.1.0...v0.2.0
[0.1.1]: https://github.com/daichirata/fluent-plugin-gcs/compare/v0.1.0...v0.1.1


test/helper.rb
require "rubygems"
require "bundler"
begin
  Bundler.setup(:default, :development)
rescue Bundler::BundlerError => e
  $stderr.puts e.message
  $stderr.puts "Run `bundle install` to install missing gems"
  exit e.status_code
end
require "test/unit"
require "timecop"

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), "..", "lib"))
$LOAD_PATH.unshift(File.dirname(__FILE__))

require "fluent/test"
require "fluent/plugin/out_gcs"

require "rr"
require "test/unit/rr"

class Test::Unit::TestCase
end

def silenced
  $stdout = StringIO.new

  yield
ensure
  $stdout = STDOUT
end


test/plugin/test_out_gcs.rb
require "helper"
require "fluent/test/driver/output"
require "fluent/test/helpers"
require "google/cloud/storage"

class GCSOutputTest < Test::Unit::TestCase
  include Fluent::Test::Helpers

  def setup
    Fluent::Test.setup
  end

  CONFIG = <<-EOC
    project test_project
    keyfile test_keyfile
    bucket test_bucket
    path log/
    <buffer>
      @type memory
      timekey_use_utc true
    </buffer>
    <system>
      log_level debug
    </system>
  EOC

  def create_driver(conf = CONFIG)
    Fluent::Test::Driver::Output.new(Fluent::Plugin::GCSOutput) do
      attr_accessor :object_creator, :encryption_opts
    end.configure(conf)
  end

  def config(*args)
    args.join("\n")
  end

  sub_test_case "configure" do
    def test_configure
      driver = create_driver
      assert_equal "test_project", driver.instance.project
      assert_equal "test_keyfile", driver.instance.keyfile
      assert_equal "test_bucket", driver.instance.bucket
      assert_equal "%{path}%{time_slice}_%{index}.%{file_extension}", driver.instance.object_key_format
      assert_equal "log/", driver.instance.path
      assert_equal :gzip, driver.instance.store_as
      assert_equal false, driver.instance.transcoding
      assert_equal true, driver.instance.auto_create_bucket
      assert_equal 4, driver.instance.hex_random_length
      assert_equal false, driver.instance.overwrite
      assert_equal "out_file", driver.instance.instance_variable_get(:@format)
      assert_equal nil, driver.instance.acl
      assert_equal nil, driver.instance.storage_class
      assert_equal nil, driver.instance.encryption_key
      assert_equal [], driver.instance.object_metadata
    end

    def test_configure_with_hex_random_length
      assert_raise Fluent::ConfigError do
        create_driver(config(CONFIG, "hex_random_length 33"))
      end
      assert_nothing_raised do
        create_driver(config(CONFIG, "hex_random_length 32"))
      end
    end

    def test_configure_with_gzip_object_creator
      driver = create_driver(config(CONFIG, "store_as gzip"))
      assert_equal true, driver.instance.object_creator.is_a?(Fluent::GCS::GZipObjectCreator)
    end

    def test_configure_with_text_object_creator
      driver = create_driver(config(CONFIG, "store_as text"))
      assert_equal true, driver.instance.object_creator.is_a?(Fluent::GCS::TextObjectCreator)
    end

    def test_configure_with_json_object_creator
      driver = create_driver(config(CONFIG, "store_as json"))
      assert_equal true, driver.instance.object_creator.is_a?(Fluent::GCS::JSONObjectCreator)
    end
  end

  def test_start
    bucket = mock!.bucket("test_bucket") { stub! }

    mock(Google::Cloud::Storage).new(
      project: "test_project",
      keyfile: "test_keyfile",
      retries: 1,
      timeout: 2,
    ) { bucket }

    driver = create_driver <<-EOC
      project test_project
      keyfile test_keyfile
      bucket test_bucket
      client_retries 1
      client_timeout 2
      <buffer>
        @type memory
        timekey_use_utc true
      </buffer>
    EOC

    driver.instance.start
  end

  def test_ensure_bucket
    bucket = stub!
    bucket.bucket { nil }
    bucket.create_bucket { "ok" }
    stub(Google::Cloud::Storage).new { bucket }

    driver = create_driver <<-EOC
      bucket test_bucket
      <buffer>
        @type memory
        timekey_use_utc true
      </buffer>
    EOC
    driver.instance.start
    assert_equal "ok", driver.instance.instance_variable_get(:@gcs_bucket)

    driver2 = create_driver <<-EOC
      bucket test_bucket
      auto_create_bucket false
      <buffer>
        @type memory
        timekey_use_utc true
      </buffer>
    EOC
    assert_raise do
      driver2.instance.start
    end
  end

  sub_test_case "foramt" do
    setup do
      bucket = stub!
      bucket.find_file { false }
      bucket.upload_file
      storage = stub!
      storage.bucket { bucket }
      stub(Google::Cloud::Storage).new { storage }

      @time = event_time("2016-01-01 12:00:00 UTC")
    end

    def test_format
      with_timezone("UTC") do
        driver = create_driver(CONFIG)
        driver.run(default_tag: "test") do
          driver.feed(@time, {"a"=>1})
          driver.feed(@time, {"a"=>2})
        end
        assert_equal %[2016-01-01T12:00:00+00:00\ttest\t{"a":1}\n], driver.formatted[0]
        assert_equal %[2016-01-01T12:00:00+00:00\ttest\t{"a":2}\n], driver.formatted[1]
      end
    end

    def test_format_included_tag_and_time
      with_timezone("UTC") do
        driver = create_driver(config(CONFIG, 'include_tag_key true', 'include_time_key true'))
        driver.run(default_tag: "test") do
          driver.feed(@time, {"a"=>1})
          driver.feed(@time, {"a"=>2})
        end
        assert_equal %[2016-01-01T12:00:00+00:00\ttest\t{"a":1,"tag":"test","time":"2016-01-01T12:00:00+00:00"}\n],
                     driver.formatted[0]
        assert_equal %[2016-01-01T12:00:00+00:00\ttest\t{"a":2,"tag":"test","time":"2016-01-01T12:00:00+00:00"}\n],
                     driver.formatted[1]
      end
    end

    def test_format_with_format_ltsv
      with_timezone("UTC") do
        driver = create_driver(config(CONFIG, 'format ltsv'))
        driver.run(default_tag: "test") do
          driver.feed(@time, {"a"=>1, "b"=>1})
          driver.feed(@time, {"a"=>2, "b"=>2})
        end
        assert_equal %[a:1\tb:1\n], driver.formatted[0]
        assert_equal %[a:2\tb:2\n], driver.formatted[1]
      end
    end

    def test_format_with_format_json
      with_timezone("UTC") do
        driver = create_driver(config(CONFIG, 'format json'))
        driver.run(default_tag: "test") do
          driver.feed(@time, {"a"=>1})
          driver.feed(@time, {"a"=>2})
        end
        assert_equal %[{"a":1}\n], driver.formatted[0]
        assert_equal %[{"a":2}\n], driver.formatted[1]
      end
    end

    def test_format_with_format_json_included_tag
      with_timezone("UTC") do
        driver = create_driver(config(CONFIG, 'format json', 'include_tag_key true'))
        driver.run(default_tag: "test") do
          driver.feed(@time, {"a"=>1})
          driver.feed(@time, {"a"=>2})
        end
        assert_equal %[{"a":1,"tag":"test"}\n], driver.formatted[0]
        assert_equal %[{"a":2,"tag":"test"}\n], driver.formatted[1]
      end
    end

    def test_format_with_format_json_included_time
      with_timezone("UTC") do
        driver = create_driver(config(CONFIG, 'format json', 'include_time_key true'))
        driver.run(default_tag: "test") do
          driver.feed(@time, {"a"=>1})
          driver.feed(@time, {"a"=>2})
        end
        assert_equal %[{"a":1,"time":"2016-01-01T12:00:00+00:00"}\n], driver.formatted[0]
        assert_equal %[{"a":2,"time":"2016-01-01T12:00:00+00:00"}\n], driver.formatted[1]
      end
    end

    def test_format_with_format_json_included_tag_and_time
      with_timezone("UTC") do
        driver = create_driver(config(CONFIG, 'format json', 'include_tag_key true', 'include_time_key true'))
        driver.run(default_tag: "test") do
          driver.feed(@time, {"a"=>1})
          driver.feed(@time, {"a"=>2})
        end
        assert_equal %[{"a":1,"tag":"test","time":"2016-01-01T12:00:00+00:00"}\n], driver.formatted[0]
        assert_equal %[{"a":2,"tag":"test","time":"2016-01-01T12:00:00+00:00"}\n], driver.formatted[1]
      end
    end
  end

  sub_test_case "write" do
    def check_upload(conf, path = nil, enc_opts = nil, upload_opts = nil, &block)
      bucket = mock!
      if block.nil?
        bucket.find_file(path, enc_opts) { false }
        bucket.upload_file(anything, path, upload_opts.merge(enc_opts))
      else
        block.call(bucket)
      end
      storage = stub!
      storage.bucket { bucket }
      stub(Google::Cloud::Storage).new { storage }

      driver = create_driver(conf)
      driver.run(default_tag: "test") do
        driver.feed(event_time("2016-01-01 15:00:00 UTC"), {"a"=>1})
      end
    end

    def test_write_with_gzip
      conf = config(CONFIG, "store_as gzip")

      enc_opts = {
        encryption_key: nil,
      }

      upload_opts = {
        metadata: {},
        acl: nil,
        storage_class: nil,
        content_type: "application/gzip",
        content_encoding: nil,
        encryption_key: nil,
      }.merge(enc_opts)

      check_upload(conf, "log/20160101_0.gz", enc_opts, upload_opts)
    end

    def test_write_with_transcoding
      conf = config(CONFIG, "store_as gzip", "transcoding true")

      enc_opts = {
        encryption_key: nil,
      }

      upload_opts = {
        metadata: {},
        acl: nil,
        storage_class: nil,
        content_type: "text/plain",
        content_encoding: "gzip",
        encryption_key: nil,
      }.merge(enc_opts)

      check_upload(conf, "log/20160101_0.gz", enc_opts, upload_opts)
    end

    def test_write_with_text
      conf = config(CONFIG, "store_as text")

      enc_opts = {
        encryption_key: nil,
      }

      upload_opts = {
        metadata: {},
        acl: nil,
        storage_class: nil,
        content_type: "text/plain",
        content_encoding: nil,
        encryption_key: nil,
      }.merge(enc_opts)

      check_upload(conf, "log/20160101_0.txt", enc_opts, upload_opts)
    end

    def test_write_with_json
      conf = config(CONFIG, "store_as json")

      enc_opts = {
        encryption_key: nil,
      }

      upload_opts = {
        metadata: {},
        acl: nil,
        storage_class: nil,
        content_type: "application/json",
        content_encoding: nil,
        encryption_key: nil,
      }.merge(enc_opts)

      check_upload(conf, "log/20160101_0.json", enc_opts, upload_opts)
    end

    def test_write_with_utc
      conf = config(CONFIG)

      enc_opts = {
        encryption_key: nil,
      }

      upload_opts = {
        metadata: {},
        acl: nil,
        storage_class: nil,
        content_type: "application/gzip",
        content_encoding: nil,
        encryption_key: nil,
      }.merge(enc_opts)

      Timecop.freeze(Time.parse("2016-01-02 01:00:00 JST")) do
        check_upload(conf, "log/20160101_0.gz", enc_opts, upload_opts)
      end
    end

    def test_write_with_placeholder_in_path
      conf = <<-CONFIG
        project test_project
        keyfile test_keyfile
        bucket test_bucket
        path log/${tag}/
        <buffer tag,time>
          @type memory
          timekey 86400
          timekey_wait 10m
          timekey_use_utc true
        </buffer>
      CONFIG

      enc_opts = {
        encryption_key: nil,
      }

      upload_opts = {
        metadata: {},
        acl: nil,
        storage_class: nil,
        content_type: "application/gzip",
        content_encoding: nil,
        encryption_key: nil,
      }.merge(enc_opts)

      Timecop.freeze(Time.parse("2016-01-02 01:00:00 JST")) do
        check_upload(conf, "log/test/20160101_0.gz", enc_opts, upload_opts)
      end
    end

    def test_write_with_encryption
      conf = config(CONFIG, "encryption_key aaa")

      enc_opts = {
        encryption_key: "aaa",
      }

      upload_opts = {
        metadata: {},
        acl: nil,
        storage_class: nil,
        content_type: "application/gzip",
        content_encoding: nil,
        encryption_key: "aaa",
      }.merge(enc_opts)

      check_upload(conf, "log/20160101_0.gz", enc_opts, upload_opts)
    end

    def test_write_with_acl
      conf = config(CONFIG, "acl auth_read")

      enc_opts = {
        encryption_key: nil,
      }

      upload_opts = {
        metadata: {},
        acl: :auth_read,
        storage_class: nil,
        content_type: "application/gzip",
        content_encoding: nil,
        encryption_key: nil,
      }.merge(enc_opts)

      check_upload(conf, "log/20160101_0.gz", enc_opts, upload_opts)
    end

    def test_write_with_storage_class
      conf = config(CONFIG, "storage_class regional")

      enc_opts = {
        encryption_key: nil,
      }

      upload_opts = {
        metadata: {},
        acl: nil,
        storage_class: :regional,
        content_type: "application/gzip",
        content_encoding: nil,
        encryption_key: nil,
      }.merge(enc_opts)

      check_upload(conf, "log/20160101_0.gz", enc_opts, upload_opts)
    end

    def test_write_with_object_metadata
      conf = config(CONFIG, <<-EOM)
        <object_metadata>
          key test-key-1
          value test-value-1
        </object_metadata>
        <object_metadata>
          key test-key-2
          value test-value-2
        </object_metadata>
      EOM

      enc_opts = {
        encryption_key: nil,
      }

      upload_opts = {
        metadata: {"test-key-1" => "test-value-1", "test-key-2" => "test-value-2"},
        acl: nil,
        storage_class: nil,
        content_type: "application/gzip",
        content_encoding: nil,
        encryption_key: nil,
      }.merge(enc_opts)

      check_upload(conf, "log/20160101_0.gz", enc_opts, upload_opts)
    end

    def test_write_with_custom_object_key_format
      conf = config(CONFIG, "object_key_format %{path}%{file_extension}/%{hex_random}/%{hostname}/%{index}/%{time_slice}/%{uuid_flush}")

      enc_opts = {
        encryption_key: nil,
      }

      upload_opts = {
        metadata: {},
        acl: nil,
        storage_class: nil,
        content_type: "application/gzip",
        content_encoding: nil,
        encryption_key: nil,
      }.merge(enc_opts)

      any_instance_of(Fluent::Plugin::Buffer::MemoryChunk) do |b|
        # Memo: Digest::MD5.hexdigest("unique_id") => "69080cee5b6d4c35a8bbf5c48335fe08"
        stub(b).unique_id { "unique_id" }
      end
      stub(SecureRandom).uuid { "uuid1" }
      stub(SecureRandom).uuid { "uuid2" }
      stub(Socket).gethostname { "test-hostname" }

      check_upload(conf) do |bucket|
        bucket.find_file(anything, enc_opts) { true }
        bucket.find_file(anything, enc_opts) { false }
        bucket.upload_file(anything, "log/gz/6908/test-hostname/1/20160101/uuid2", upload_opts.merge(enc_opts))
      end
    end

    def test_write_with_overwrite_true
      conf = config(CONFIG, "object_key_format %{path}%{time_slice}.%{file_extension}", "overwrite true")

      enc_opts = {
        encryption_key: nil,
      }

      upload_opts = {
        metadata: {},
        acl: nil,
        storage_class: nil,
        content_type: "application/gzip",
        content_encoding: nil,
        encryption_key: nil,
      }.merge(enc_opts)

      check_upload(conf) do |bucket|
        bucket.find_file(anything, enc_opts) { true }
        bucket.find_file(anything, enc_opts) { true }
        bucket.upload_file(anything, "log/20160101.gz", upload_opts.merge(enc_opts))
      end
    end

    def test_write_with_overwrite_false
      conf = config(CONFIG, "object_key_format %{path}%{time_slice}.%{file_extension}", "overwrite false")

      enc_opts = {
        encryption_key: nil,
      }

      assert_raise do
        silenced do
          check_upload(conf) do |bucket|
            bucket.find_file(anything, enc_opts) { true }
            bucket.find_file(anything, enc_opts) { true }
          end
        end
      end
    end
  end
end


lib/fluent/plugin/gcs/object_creator.rb
require "tempfile"
require "zlib"
require "lzo"

module Fluent
  module GCS
    def self.discovered_object_creator(store_as, transcoding: nil)
      case store_as
      when :gzip
        Fluent::GCS::GZipObjectCreator.new(transcoding)
      when :json
        Fluent::GCS::JSONObjectCreator.new
      when :text
        Fluent::GCS::TextObjectCreator.new
      when :lzo
        Fluent::GCS::LzoObjectCreator.new(transcoding)
      end
    end

    class ObjectCreator
      def content_type
        raise NotImplementedError
      end

      def content_encoding
        nil
      end

      def file_extension
        raise NotImplementedError
      end

      def write(chunk, io)
        raise NotImplementedError
      end

      def create(chunk, &block)
        Tempfile.create("fluent-plugin-gcs") do |f|
          f.binmode
          f.sync = true
          write(chunk, f)
          block.call(f)
        end
      end
    end

    class GZipObjectCreator < ObjectCreator
      def initialize(transcoding)
        @transcoding = transcoding
      end

      def content_type
        @transcoding ? "text/plain" : "application/gzip"
      end

      def content_encoding
        @transcoding ? "gzip" : nil
      end

      def file_extension
        "gz"
      end

      def write(chunk, io)
        writer = Zlib::GzipWriter.new(io)
        chunk.write_to(writer)
        writer.finish
      end
    end

    class LzoObjectCreator < ObjectCreator
      def initialize(transcoding)
        @transcoding = transcoding
      end

      def content_type
        @transcoding ? "text/plain" : "application/x-lzo"
      end

      def content_encoding
        @transcoding ? "lzop" : nil
      end

      def file_extension
        "lzo"
      end

      def write(chunk, io)
        writer = LZO::LzopCompressor.new io, name: '', mode: 0100644, mtime: Time.now
        chunk.write_to(writer)
        writer.close # need to close, as LzopCompressor writes LZO end symbols
      end
    end

    class TextObjectCreator < ObjectCreator
      def content_type
        "text/plain"
      end

      def file_extension
        "txt"
      end

      def write(chunk, io)
        chunk.write_to(io)
      end
    end

    class JSONObjectCreator < TextObjectCreator
      def content_type
        "application/json"
      end

      def file_extension
        "json"
      end
    end
  end
end


lib/fluent/plugin/out_gcs.rb
require "digest/md5"
require "securerandom"
require "socket"

require "fluent/plugin/gcs/object_creator"
require "fluent/plugin/gcs/version"
require "fluent/plugin/output"

module Fluent::Plugin
  class GCSOutput < Output
    Fluent::Plugin.register_output("gcs", self)

    helpers :compat_parameters, :formatter, :inject

    def initialize
      super
      require "google/cloud/storage"
    end

    config_param :project, :string,  default: nil,
                 desc: "Project identifier for GCS"
    config_param :keyfile, :string, default: nil,
                 desc: "Path of GCS service account credentials JSON file"
    config_param :client_retries, :integer, default: nil,
                 desc: "Number of times to retry requests on server error"
    config_param :client_timeout, :integer, default: nil,
                 desc: "Default timeout to use in requests"
    config_param :bucket, :string,
                 desc: "Name of a GCS bucket"
    config_param :object_key_format, :string, default: "%{path}%{time_slice}_%{index}.%{file_extension}",
                 desc: "Format of GCS object keys"
    config_param :path, :string, default: "",
                 desc: "Path prefix of the files on GCS"
    config_param :store_as, :enum, list: %i(gzip json text), default: :gzip,
                 desc: "Archive format on GCS"
    config_param :transcoding, :bool, default: false,
                 desc: "Enable the decompressive form of transcoding"
    config_param :auto_create_bucket, :bool, default: true,
                 desc: "Create GCS bucket if it does not exists"
    config_param :hex_random_length, :integer, default: 4,
                 desc: "Max length of `%{hex_random}` placeholder(4-16)"
    config_param :overwrite, :bool, default: false,
                 desc: "Overwrite already existing path"
    config_param :format, :string, default: "out_file",
                 desc: "Change one line format in the GCS object"
    config_param :acl, :enum, list: %i(auth_read owner_full owner_read private project_private public_read), default: nil,
                 desc: "Permission for the object in GCS"
    config_param :storage_class, :enum, list: %i(dra nearline coldline multi_regional regional standard), default: nil,
                 desc: "Storage class of the file"
    config_param :encryption_key, :string, default: nil, secret: true,
                 desc: "Customer-supplied, AES-256 encryption key"
    config_param :blind_write, :bool, default: false,
                 desc: "Whether to check if object already exists by given GCS path. Allows avoiding giving storage.object.get permission"
    config_section :object_metadata, required: false do
      config_param :key, :string, default: ""
      config_param :value, :string, default: ""
    end

    DEFAULT_FORMAT_TYPE = "out_file"

    config_section :format do
      config_set_default :@type, DEFAULT_FORMAT_TYPE
    end

    config_section :buffer do
      config_set_default :chunk_keys, ['time']
      config_set_default :timekey, (60 * 60 * 24)
    end

    MAX_HEX_RANDOM_LENGTH = 32

    def configure(conf)
      compat_parameters_convert(conf, :buffer, :formatter, :inject)
      super

      if @hex_random_length > MAX_HEX_RANDOM_LENGTH
        raise Fluent::ConfigError, "hex_random_length parameter should be set to #{MAX_HEX_RANDOM_LENGTH} characters or less."
      end

      # The customer-supplied, AES-256 encryption key that will be used to encrypt the file.
      @encryption_opts = {
        encryption_key: @encryption_key,
      }

      if @object_metadata
        @object_metadata_hash = @object_metadata.map {|m| [m.key, m.value] }.to_h
      end

      @formatter = formatter_create

      @object_creator = Fluent::GCS.discovered_object_creator(@store_as, transcoding: @transcoding)
      # For backward compatibility
      # TODO: Remove time_slice_format when end of support compat_parameters
      @configured_time_slice_format = conf['time_slice_format']
      @time_slice_with_tz = Fluent::Timezone.formatter(@timekey_zone, @configured_time_slice_format || timekey_to_timeformat(@buffer_config['timekey']))
    end

    def start
      @gcs = Google::Cloud::Storage.new(
        project: @project,
        keyfile: @keyfile,
        retries: @client_retries,
        timeout: @client_timeout
      )
      @gcs_bucket = @gcs.bucket(@bucket)

      ensure_bucket
      super
    end

    def format(tag, time, record)
      r = inject_values_to_record(tag, time, record)
      @formatter.format(tag, time, r)
    end

    def multi_workers_ready?
      true
    end

    def write(chunk)
      path = generate_path(chunk)

      @object_creator.create(chunk) do |obj|
        opts = {
          metadata: @object_metadata_hash,
          acl: @acl,
          storage_class: @storage_class,
          content_type: @object_creator.content_type,
          content_encoding: @object_creator.content_encoding,
        }
        opts.merge!(@encryption_opts)

        log.debug { "out_gcs: upload chunk:#{chunk.key} to gcs://#{@bucket}/#{path} options: #{opts}" }
        @gcs_bucket.upload_file(obj.path, path, **opts)
      end
    end

    private

    def ensure_bucket
      return unless @gcs_bucket.nil?

      if !@auto_create_bucket
        raise "bucket `#{@bucket}` does not exist"
      end
      log.info "creating bucket `#{@bucket}`"
      @gcs_bucket = @gcs.create_bucket(@bucket)
    end

    def hex_random(chunk)
      Digest::MD5.hexdigest(chunk.unique_id)[0...@hex_random_length]
    end

    def check_object_exists(path)
      if !@blind_write
        return @gcs_bucket.find_file(path, **@encryption_opts)
      else
        return false
      end
    end

    def generate_path(chunk, i = 0, prev = nil)
      metadata = chunk.metadata
      time_slice = if metadata.timekey.nil?
                     ''.freeze
                   else
                     @time_slice_with_tz.call(metadata.timekey)
                   end
      tags = {
        "%{file_extension}" => @object_creator.file_extension,
        "%{hex_random}" => hex_random(chunk),
        "%{hostname}" => Socket.gethostname,
        "%{index}" => i,
        "%{path}" => @path,
        "%{time_slice}" => time_slice,
        "%{uuid_flush}" => SecureRandom.uuid,
      }
      path = @object_key_format.gsub(Regexp.union(tags.keys), tags)
      path = extract_placeholders(path, chunk)
      return path unless check_object_exists(path)

      if path == prev
        if @overwrite
          log.warn "object `#{path}` already exists but overwrites it"
          return path
        end
        raise "object `#{path}` already exists"
      end
      generate_path(chunk, i + 1, path)
    end

    # This is stolen from Fluentd
    def timekey_to_timeformat(timekey)
      case timekey
      when nil          then ''
      when 0...60       then '%Y%m%d%H%M%S' # 60 exclusive
      when 60...3600    then '%Y%m%d%H%M'
      when 3600...86400 then '%Y%m%d%H'
      else                   '%Y%m%d'
      end
    end
  end
end
